support ARM and ARM64 architecture?

as per title, any chance to add these architectures?
I have tried building from Dockerfile but it fails on step 6:
`COPY go.mod go.sum ./` displaying no such file or directory 
plus do provide docker-compose file

As discussed on slack with @ldez I've wipped up a strawman for building multiarch images and bundling inside a manifest. It basically builds and pushes images for all platforms and in a second step bundles and publishes a manifest. 

I've tested this on OSX which but- since the runtime image is only `COPYied` into and does not use `RUN`- this build process should work on any platform. Currently the script uses `jq` which could be replaced by `docker run realguess/jq jq`.

````shell
#!/bin/bash
set -e

# base docker image tag
TAG="andig/traefik-certs-dumper"

# only linux for now
OS=linux

# target platforms
declare -a PLATFORMS=( "amd64" "arm64" "arm.v6" )

# images from Dockerfile
RUNTIME_IMAGE=$(grep "FROM alpine" < Dockerfile | sed "s/FROM //")

# platform-specific docker file
DOCKER_TEMP=/tmp/Dockerfile.temp

# cache the docker manifest
function manifest () {
    local IMAGE=$1
    local MANIFEST_FILE=/tmp/manifest.$IMAGE.json

    if [ ! -f "$MANIFEST_FILE" ]; then
        docker pull "$IMAGE"
        docker manifest inspect "$IMAGE" > "$MANIFEST_FILE"
    fi
}

# get platform image hash from docker manifest
function hash () {
    local IMAGE=$1
    local MANIFEST_FILE=/tmp/manifest.$IMAGE.json

    local ARCHITECTURE="$2"
    local VARIANT="$3"
    local HASH

    if [ -z "$VARIANT" ]; then
        HASH=$(jq -r ".manifests[] | select(.platform.architecture == \"$ARCHITECTURE\") | .digest" < "$MANIFEST_FILE")
    else
        HASH=$(jq -r ".manifests[] | select(.platform.architecture == \"$ARCHITECTURE\" and .platform.variant == \"$VARIANT\") | .digest" < "$MANIFEST_FILE")
    fi

    echo "$HASH"
}

# https://stackoverflow.com/questions/1527049/how-can-i-join-elements-of-an-array-in-bash
function taglist {
    printf "$TAG:latest-%s " "${@/#/}"
}

# get manifests
manifest "$RUNTIME_IMAGE"

# main
for platform in "${PLATFORMS[@]}"; do 
    # split architecture.version
    IFS='.' read -ra p <<< "$platform"

    # add xargs to trim whitespace
    RUNTIME_HASH=$(hash "$RUNTIME_IMAGE" "${p[0]}" "${p[1]}")

    # target architecture
    GOARCH=${p[0]}

    # create temp Dockerfile for target architecture and build
    sed "s/RUN make/RUN GOARCH=$GOARCH make/" Dockerfile | sed "s/$RUNTIME_IMAGE/$RUNTIME_IMAGE@$RUNTIME_HASH/" > "$DOCKER_TEMP"
    docker build -t "$TAG:latest-$platform" -f $DOCKER_TEMP .
    rm "$DOCKER_TEMP"
done

# push images
for platform in "${PLATFORMS[@]}"; do 
    docker push "$TAG:latest-$platform"
done

# create manifest
TAG_LIST=$(taglist "${PLATFORMS[@]}")
# shellcheck disable=SC2086
docker manifest create --amend "$TAG:latest" $TAG_LIST

for platform in "${PLATFORMS[@]}"; do 
    # split architecture.version
    IFS='.' read -ra p <<< "$platform"

    ARCHITECTURE=${p[0]}
    VARIANT=${p[1]}
    
    # docker and go architectures don't match
    if [ "arm" == "$ARCHITECTURE" ] && [ ! -z "$VARIANT" ]; then
        VARIANT=$ARCHITECTURE$VARIANT
    fi

    if [ -z "$VARIANT" ]; then
        docker manifest annotate "$TAG:latest" "$TAG:latest-$platform" --os "$OS" --arch "$ARCHITECTURE"
    else
        docker manifest annotate "$TAG:latest" "$TAG:latest-$platform" --os "$OS" --arch "$ARCHITECTURE" --variant "$VARIANT"
    fi
done

docker manifest push "$TAG:latest"
````

The result is available as `andig/traefik-certs-dumper:latest` for demo purposes. This is intentionally not a PR as I didn't know how to integrate it with the `goreleaser`.

I imagine one would run `build-docker.sh` inside travis `after_sucess` like this:

````
env:
  global:
    - PATH=/home/travis/gopath/bin:$PATH
    # docker credentials
    - secure: "MMDlT1..."
    - secure: "aFjBMVNpdv..."
    - COMMIT=${TRAVIS_COMMIT::8}

after_success:
  - docker login -u $DOCKER_USER -p $DOCKER_PASS
  - export REPO=andig/gravo
  - docker pull $REPO:latest
  - |
    if [ "$TRAVIS_BRANCH" == "master" ]; then 
      ./build-docker.sh
    fi
````

docker credentials are added to travis using the travis cli or through travis ui.

--- a/.dockerignore
+++ b/.dockerignore
@@ -6,3 +6,4 @@ dumpcerts.sh
 acme.json
 acme-backup.json
 traefik-certs-dumper
+build-docker.sh

--- a/.travis.yml
+++ b/.travis.yml
@@ -30,3 +30,9 @@ deploy:
     on:
       tags: true
       condition: $TRAVIS_GO_VERSION =~ ^1\.x$
+  - provider: script
+    skip_cleanup: true
+    script: make publish-images
+    on:
+      tags: true
+      condition: $TRAVIS_GO_VERSION =~ ^1\.x$

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,7 +4,8 @@ go:
   - 1.11.x
   - 1.x
 
-sudo: false
+services:
+  - docker
 
 env:
   - GO111MODULE=on
@@ -22,6 +23,13 @@ before_install:
 install:
   - echo "TRAVIS_GO_VERSION=$TRAVIS_GO_VERSION"
   - go mod download
+
+before_deploy:
+  - >
+    if ! [ "$BEFORE_DEPLOY_RUN" ]; then
+      export BEFORE_DEPLOY_RUN=1;
+      echo "${DOCKER_PASSWORD}" | docker login -u "${DOCKER_USERNAME}" --password-stdin
+    fi
 
 deploy:
   - provider: script

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,5 +1,10 @@
 FROM golang:1-alpine as builder
 
+ARG RUNTIME_HASH
+ARG GOARCH
+ARG GOARM
+ARG GOOS
+
 RUN apk --update upgrade \
     && apk --no-cache --no-progress add git make gcc musl-dev
 
@@ -10,9 +15,9 @@ COPY go.mod go.sum ./
 RUN go mod download
 
 COPY . .
-RUN make build
+RUN GOARCH=${GOARCH} GOARM=${GOARM} GOOS=${GOOS} make build
 
-FROM alpine:3.9
+FROM alpine:3.9${RUNTIME_HASH}
 RUN apk --update upgrade \
     && apk --no-cache --no-progress add ca-certificates
 

--- a/Makefile
+++ b/Makefile
@@ -16,7 +16,11 @@ clean:
 
 build: clean
 	@echo Version: $(VERSION) $(BUILD_DATE)
-	go build -v -ldflags '-X "github.com/ldez/traefik-certs-dumper/cmd.version=${VERSION}" -X "github.com/ldez/traefik-certs-dumper/cmd.commit=${SHA}" -X "github.com/ldez/traefik-certs-dumper/cmd.date=${BUILD_DATE}"'
+	go build -v -ldflags '-X "github.com/ldez/traefik-certs-dumper/cmd.version=${VERSION}" -X "github.com/ldez/traefik-certs-dumper/cmd.commit=${SHA}" -X "github.com/ldez/traefik-certs-dumper/cmd.date=${BUILD_DATE}"' -o traefik-certs-dumper
 
 checks:
 	golangci-lint run
+
+publish-images:
+	VERSION=$(TAG_NAME) ./build-docker.sh
+	VERSION="latest" ./build-docker.sh

--- a/build-docker.sh
+++ b/build-docker.sh
@@ -11,7 +11,7 @@ OS=linux
 declare -a PLATFORMS=( "amd64" "arm.v6" )
 
 # images from Dockerfile
-IMAGE=$(grep "FROM alpine" < Dockerfile | sed "s/FROM //" | sed 's/\$.*//')
+IMAGE=$(grep "{RUNTIME_HASH}" < Dockerfile | sed "s/FROM //" | sed 's/\$.*//')
 
 # manifest cache file
 MANIFEST_FILE=/tmp/manifest.$IMAGE.json

--- a/build-docker.sh
+++ b/build-docker.sh
@@ -1,0 +1,76 @@
+#!/bin/bash
+set -ex
+
+# base docker image tag
+TAG="andig/traefik-certs-dumper"
+
+# only linux for now
+OS=linux
+
+# target platforms in docker manifest notation
+declare -a PLATFORMS=( "amd64" "arm.v6" )
+
+# images from Dockerfile
+IMAGE=$(grep "FROM alpine" < Dockerfile | sed "s/FROM //" | sed 's/\$.*//')
+
+# manifest cache file
+MANIFEST_FILE=/tmp/manifest.$IMAGE.json
+
+# get platform image hash from docker manifest
+function hash () {
+    local ARCHITECTURE VARIANT HASH
+    read -r ARCHITECTURE VARIANT <<< "$@"
+
+    if [ -z "$VARIANT" ]; then
+        HASH=$(jq -r ".manifests[] | select(.platform.architecture == \"$ARCHITECTURE\") | .digest" < "$MANIFEST_FILE")
+    else
+        HASH=$(jq -r ".manifests[] | select(.platform.architecture == \"$ARCHITECTURE\" and .platform.variant == \"$VARIANT\") | .digest" < "$MANIFEST_FILE")
+    fi
+
+    echo "$HASH"
+}
+
+# get manifest
+if [ ! -f "$MANIFEST_FILE" ]; then
+    docker pull "$IMAGE"
+    docker manifest inspect "$IMAGE" > "$MANIFEST_FILE"
+fi
+
+# main
+for platform in "${PLATFORMS[@]}"; do 
+    # split architecture.version
+    IFS='.' read -r ARCHITECTURE VARIANT <<< "$platform"
+
+    # add xargs to trim whitespace
+    RUNTIME_HASH=$(hash "$ARCHITECTURE" "$VARIANT")
+
+    # arm architectures flavors, strip "v" prefix
+    GOARM=${VARIANT:1}
+
+    # build for target runtime image and architecture
+    docker build --build-arg RUNTIME_HASH=@${RUNTIME_HASH} --build-arg GOARCH=${ARCHITECTURE} --build-arg GOARM=${GOARM} -t "$TAG:latest-$platform" .
+done
+
+# push images
+for platform in "${PLATFORMS[@]}"; do 
+    docker push "$TAG:latest-$platform"
+done
+
+# create manifest
+TAG_LIST=$(printf "$TAG:latest-%s " "${PLATFORMS[@]}")
+# shellcheck disable=SC2086
+docker manifest create --amend "$TAG:latest" $TAG_LIST
+
+for platform in "${PLATFORMS[@]}"; do 
+    # split architecture.version
+    IFS='.' read -r ARCHITECTURE VARIANT <<< "$platform"
+    
+    # docker and go architectures don't match
+    if [ "arm" == "$ARCHITECTURE" ] && [ -n "$VARIANT" ]; then
+        VARIANT="$ARCHITECTURE$VARIANT"
+    fi
+
+    docker manifest annotate "$TAG:latest" "$TAG:latest-$platform" --os "$OS" --arch "$ARCHITECTURE" --variant "$VARIANT"
+done
+
+docker manifest push "$TAG:latest"

--- a/build-docker.sh
+++ b/build-docker.sh
@@ -1,23 +1,33 @@
-#!/bin/bash
-set -ex
+#!/usr/bin/env bash
 
-# base docker image tag
-TAG="andig/traefik-certs-dumper"
+set -o errexit
+set -o pipefail
+
+# safe guard
+if [ -n "$TRAVIS_TAG" ] && [ -n "$VERSION" ]; then
+  echo "Deploying..."
+else
+  echo "Skipping deploy"
+  exit 0
+fi
+
+# base docker image name
+IMAGE_NAME="ldez/traefik-certs-dumper"
 
 # only linux for now
 OS=linux
 
 # target platforms in docker manifest notation
-declare -a PLATFORMS=( "amd64" "arm.v6" )
+declare -a PLATFORMS=( "amd64" "arm.v6" "arm.v7")
 
 # images from Dockerfile
-IMAGE=$(grep "{RUNTIME_HASH}" < Dockerfile | sed "s/FROM //" | sed 's/\$.*//')
+FROM_IMAGE=$(grep "{RUNTIME_HASH}" < Dockerfile | sed "s/FROM //" | sed 's/\$.*//')
 
 # manifest cache file
-MANIFEST_FILE=/tmp/manifest.$IMAGE.json
+MANIFEST_FILE=/tmp/tcd-manifest.${FROM_IMAGE}.json
 
 # get platform image hash from docker manifest
-function hash () {
+function platformHash () {
     local ARCHITECTURE VARIANT HASH
     read -r ARCHITECTURE VARIANT <<< "$@"
 
@@ -32,45 +42,44 @@ function hash () {
 
 # get manifest
 if [ ! -f "$MANIFEST_FILE" ]; then
-    docker pull "$IMAGE"
-    docker manifest inspect "$IMAGE" > "$MANIFEST_FILE"
+    docker pull "$FROM_IMAGE"
+    DOCKER_CLI_EXPERIMENTAL=enabled docker manifest inspect "$FROM_IMAGE" > "$MANIFEST_FILE"
 fi
 
-# main
-for platform in "${PLATFORMS[@]}"; do 
+# create and push images
+for platform in "${PLATFORMS[@]}"; do
     # split architecture.version
     IFS='.' read -r ARCHITECTURE VARIANT <<< "$platform"
 
     # add xargs to trim whitespace
-    RUNTIME_HASH=$(hash "$ARCHITECTURE" "$VARIANT")
+    RUNTIME_HASH=$(platformHash "$ARCHITECTURE" "$VARIANT")
 
     # arm architectures flavors, strip "v" prefix
     GOARM=${VARIANT:1}
 
     # build for target runtime image and architecture
-    docker build --build-arg RUNTIME_HASH=@${RUNTIME_HASH} --build-arg GOARCH=${ARCHITECTURE} --build-arg GOARM=${GOARM} -t "$TAG:latest-$platform" .
-done
+    docker build --build-arg="RUNTIME_HASH=@${RUNTIME_HASH}" --build-arg="GOARCH=${ARCHITECTURE}" --build-arg="GOARM=${GOARM}" -t "$IMAGE_NAME:${VERSION}-$platform" .
 
-# push images
-for platform in "${PLATFORMS[@]}"; do 
-    docker push "$TAG:latest-$platform"
+    # push images
+    docker push "$IMAGE_NAME:${VERSION}-$platform"
 done
 
 # create manifest
-TAG_LIST=$(printf "$TAG:latest-%s " "${PLATFORMS[@]}")
+TAG_LIST=$(printf "$IMAGE_NAME:${VERSION}-%s " "${PLATFORMS[@]}")
 # shellcheck disable=SC2086
-docker manifest create --amend "$TAG:latest" $TAG_LIST
+DOCKER_CLI_EXPERIMENTAL=enabled docker manifest create --amend "$IMAGE_NAME:${VERSION}" $TAG_LIST
 
-for platform in "${PLATFORMS[@]}"; do 
+for platform in "${PLATFORMS[@]}"; do
     # split architecture.version
     IFS='.' read -r ARCHITECTURE VARIANT <<< "$platform"
-    
+
     # docker and go architectures don't match
     if [ "arm" == "$ARCHITECTURE" ] && [ -n "$VARIANT" ]; then
         VARIANT="$ARCHITECTURE$VARIANT"
     fi
 
-    docker manifest annotate "$TAG:latest" "$TAG:latest-$platform" --os "$OS" --arch "$ARCHITECTURE" --variant "$VARIANT"
+    DOCKER_CLI_EXPERIMENTAL=enabled docker manifest annotate "$IMAGE_NAME:${VERSION}" "$IMAGE_NAME:${VERSION}-$platform" --os "$OS" --arch "$ARCHITECTURE" --variant "$VARIANT"
 done
 
-docker manifest push "$TAG:latest"
+# push manifest
+DOCKER_CLI_EXPERIMENTAL=enabled docker manifest push "$IMAGE_NAME:${VERSION}"

